estimagic/visualization/comparison_plot.py

```diff
@@ -0,0 +1,440 @@
+"""
+Draw an interactive comparison plot of named result dictionaries.
+
+The plot can plot many results for large numbers of parameters
+against each other.
+
+The plot can answer the following questions:
+
+1. How are the parameters distributed?
+
+2. How large are the differences in parameter estimates between results
+    compared to the uncertainty around the parameter estimates?
+
+3. Are parameters of groups of results clustered?
+
+"""
+import numpy as np
+import pandas as pd
+from bokeh.layouts import gridplot
+from bokeh.models import BoxSelectTool
+from bokeh.models import ColumnDataSource
+from bokeh.models import HoverTool
+from bokeh.models import TapTool
+from bokeh.models import Title
+from bokeh.models.callbacks import CustomJS
+from bokeh.models.widgets import CheckboxGroup
+from bokeh.plotting import figure
+from bokeh.plotting import show
+
+from estimagic.optimization.utilities import index_element_to_string
+
+
+def comparison_plot(
+    res_dict, color_dict=None, height=None, width=600, point_estimate_plot_kwargs=None
+):
+    """Make a comparison plot from a dictionary containing optimization results.
+
+    Args:
+        res_dict (dict): The keys are the names of different models.
```

Comment:
What is the difference between the keys of ``res_dict`` and ``model_class``?

Reply:
``model_class`` is a string, not a dictionary. 

The inputs of the comparison_plot might be be improved.

At the moment the ``res_dict`` is a nested dictionary with the model_name (which must be unique) as key and a dictionary as value. That inner dictionary at the moment has at most two entries: Firstly, ``result_df`` which maps to the params DataFrame returned by ``estimagic.optimization.maximize`` or ``minimize``. Secondly ``model_class`` which maps just to a string (that should be the same for more than one model_name). 

I have considered just having the ``model_class`` as an optional column in the ``result_df`` to have a flat dictionary as input. Then the ``res_dict`` would be flat and a little more user-friendly but it seems superfluous and unelegant to have a column with a hundred times or more the same string for every of possibly many DataFrames passed to comparison_plot. 

I think @janosg  and I talked about this trade-off some time ago and decided on the nested structure but if it is confusing we might want to reconsider, especially as it turns out that we really just have that single additional entry in the inner dictionary.

Reply:
Thanks for the explanation. But, I think I was not clear. What is the difference between ``model_class`` and ``model_name``? For me it seems, that we would be fine with just one name for a set of parameters. Why do you need a second level?

If we only need one name, the input to comparison plot just has to be an iterable. Then, if elements of the iterable are of length 2, take the first as the model name and the second as the model's parameters. Thus, you can pass a list of tuples ``[("model_a", params_a), ("model_b", params_b)]`` as well as a dictionary. If the elements are of length 1, number the models and use that as the name.

Having the nested structure does not seem intuitive to me, but I might be wrong as you probably have thought about more use-cases than I have.

Reply:
``model_name`` uniquely identifies one model, in this case one optimization result. ``model_class`` allows you to group models together. For example in the opening message L-BFGS-B, SLSQP and TNC (each run several times from different starting values) are compared. Other use cases could be having different sets of control variables, instrumenting a variable of interest or not, comparing ML vs GMM ... 

I like your idea of having a list of tuples instead of the nested dictionary. That would make it much easier to use in case a user has no model_classes nor valuable model_names!

Reply:
Now, I understand. What about a combination?

```
parameters = {
    "params_1": [
        ("TNC", params_1_tnc),
        ("L-BFGS-B", "params_1_lbfgsb")
    ],
    "params_2": [
        ("TNC", params_2_tnc),
        ("L-BFGS-B", "params_2_lbfgsb")
    ],
}
```
Or what do you think are the chances of having a plot where you are able to flip the dimensions. I mean not only sorting by ``params_1`` first, but by ``TNC`` and ``L-BFGS-B``? Then, as there is no predefined hierarchy anymore, you could do
```
parameters = [
    ("params_1", "TNC", params_1_tnv),
    ("params_2", "TNC", params_2_tnc),
    ...
]
```
where dimensions are everything but the last tuple entry. I think this would be pretty easy to aggregate just with for loops and zip. Easier than the dictionary. Also it would not get deeply nested in case you add other dimensions.